fastly/resource_fastly_domain_v1.go

```
@@ -0,0 +1,122 @@
+package fastly
+
+import (
+	"context"
+	"log"
+
+	gofastly "github.com/fastly/go-fastly/v9/fastly"
+	v1 "github.com/fastly/go-fastly/v9/fastly/domains/v1"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+)
+
+func resourceFastlyDomainV1() *schema.Resource {
+	return &schema.Resource{
+		CreateContext: resourceFastlyDomainV1Create,
+		ReadContext:   resourceFastlyDomainV1Read,
+		UpdateContext: resourceFastlyDomainV1Update,
+		DeleteContext: resourceFastlyDomainV1Delete,
+		Importer: &schema.ResourceImporter{
+			StateContext: schema.ImportStatePassthroughContext,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"domain_id": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "The Domain Identifier (UUID).",
+			},
+			"fqdn": {
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+				Description: "The fully-qualified domain name for your domain. Can be created, but not updated.",
```

Generally, FQDNs consist the trailing `.` so it will be `www.example.com.` instead of `www.example.com`. We have at least used FQDN to mean the latter, without the trailing `.`. It made me wonder if that has caused confusion in the past. I think UDM uses the same domain normalization that other parts of spotless uses so we likely strip it off even if the customer used proper FQDNs but it did made me wonder.

---

Yeah, in most cases we should not require the trailing '.', but we willing to accept it if is present. However, that brings up an important point for the TF provider: if the user specifies the FQDN with the trailing '.', it *must* be preserved by the API and returned exactly the way the user entered it. Otherwise the provider will think that the resource has been changed since it was created, and with `ForceNew: true` it will attempt to destroy the resource and recreate it.

In other words: the API must silently accept and return names in both formats, even if it only uses one format internally, or it must noisily reject names in the non-preferred format.

---

This is a great point @kpfleming. Thanks.

If the use gives `www.example.com.` and the API happily accepts it but internally normalises it to strip the trailing dot, then when we _READ_ from the API to check the Terraform state against the user's Terraform config file, then there will always be a diff shown to the user as the state file will have `www.example.com` and not `www.example.com.` like in the user's config and it'll presume a change occurred, when in practice it isn't actually a different behaviour.

So one of the following two options has to happen, either... 

1. The TF provider documentation explicitly state what it expects (trailing dot or no dot), and that _must_ correlate to what the API actually returns (which I think is "no trailing dot" but I could be wrong).
2. Or, the user enters whatever they like (trailing dot or not dot, doesn't matter) but the API _must_ return that same value and not a normalised value where it's added or removed.

@tundal45 @bryangm @gschorkopf can y'all advise. Thanks!

---

@Integralist I would tend to lean to option 1 at this time.

While we only recently exposed `/domains/v1` publicly as part of the UDM project, our `/tls/subscription` and `/tls/certificates` endpoints have used Spotless's domain logic internally to manage domains that appear as SANs on certificates Fastly uses for quite some time. 

While I don't disagree that option 2 would be the _best_ option, the effort and prioritization of the work would take some time and option 1 could clear up the issue considerably sooner.

---

On a functional note, as perhaps pointed out already: Spotless always gets rid of the trailing `.` in the `domains.fqdn` DB column. We append it in some cases when interacting with APIs that require it (GlobalSign APIs for example).